questionGenerator.py

```diff
@@ -0,0 +1,318 @@
+        overwrite (bool, optional): Whether to overwrite existing question data. Defaults to False.
+
+    Returns:
+        QuestionData: The question data object.
+    """
+    if not config.overwriteQuestionData and not overwrite:
+        questionData = QuestionData(config, videoData, topicModeller, load=True)
+
+        if questionData.responseData is not None:
+            printAndLog("Question Data loaded from saved files.")
+            printAndLog(
+                f"Question Data Count: {len(questionData.responseData)}",
+                logOnly=True,
+            )
+            return questionData
+
+    printAndLog("Generating Question Data...", logOnly=True)
+
+    questionData = QuestionData(config, videoData, topicModeller, load=False)
+    questionData.saveQuestionData()
+
+    printAndLog("Question Data generated and saved for current configuration.")
+    printAndLog(
+        f"Question Data Count: {len(questionData.responseData)}",
+        logOnly=True,
+    )
+    return questionData
+
+
+def getClusteredTopics(videoData, topicModeller):
+    """
+    Get clustered topics from videoData using topicModeller.
+
+    Args:
+        videoData (object): The video data object.
+        topicModeller (object): The topic modeller object.
+
+    Returns:
+        pandas.DataFrame: A DataFrame containing the clustered topics with the following columns:
+            - Topic: The dominant topic ID.
+            - Words: The words associated with the topic.
+            - Frequency: The frequency of the topic.
+            - Timestamp: The timestamp of the topic.
+            - Start: The start time of the topic segment.
+            - End: The end time of the topic segment.
+            - Topic Title: The title of the topic.
+
+    """
+    # Sort the topic by timsetamp and frequency.
+    sortedTopics = topicModeller.topicsOverTime.sort_values(
+        ["Timestamp", "Frequency"], ascending=False
+    )
+
+    # Remove any segments where the topic is not relevant.
+    # These are segments where the topic ID is -1.
+    filteredGroups = []
+    for group in sortedTopics.groupby("Timestamp"):
+        dominantTopic = group[1].iloc[0]["Topic"]
+        if dominantTopic != -1:
+            filteredGroups.append(group[1])
+
+    filteredGroupsDF = pd.concat(filteredGroups)
+
+    # Clean up the words column.
+    filteredGroupsDF["Words"] = filteredGroupsDF["Words"].apply(
+        lambda keyPhrase: keyPhrase.rstrip(", ")
+    )
+
+    # We also filter out any topics that are not relevant in other segments.
+    filteredTopics = filteredGroupsDF[filteredGroupsDF["Topic"] != -1]
+
+    # We can merge any segments where the main topic is the same.
+    # This gives the inutition that the same topic is being discussed over a period of time.
+    clusteredTopics = (
+        filteredTopics.groupby(
+            (filteredTopics["Topic"] != filteredTopics["Topic"].shift()).cumsum()
+        )
+        .agg(
```

Comment:
This function implementation is raising error. So Understand if that error is wrap in `try-catch`

Reply:
Do you have any more info on what video caused this error? I will wrap the function calls in a `try-catch` but I'd also like to replicate the error itself, it might indicate something needs to be adjusted for topic extraction.

Reply:
New Quizzes Video